### PR TITLE
Add ray rdd

### PR DESCRIPTION
**Pull request by yangw1234**

### CODE REVIEW

**Comment by yangw1234** [comment]
jenkins passed http://172.168.2.101:8080/view/ZOO-PR/job/ZOO-PR-Validation/4421/

**Comment by yangw1234** [comment]
http://172.168.2.101:8080/view/ZOO-PR/job/ZOO-PR-Validation/4425/

File: pyzoo/zoo/orca/data/ray_rdd.py [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+        return 0
+
+    def get_partition_ref(self, idx):
+        return self.partitions[idx]
+
+    def get_multiple_partition_refs(self, idxs):
+        return [self.partitions[idx] for idx in idxs]
+
+    def get_all_partition_refs(self):
+        return [self.partitions[i] for i in range(len(self.partitions))]
```
Comment (hkvision) [inline pyzoo/zoo/orca/data/ray_rdd.py:40]: why not directly return self.partitions?

Comment (yangw1234) [reply pyzoo/zoo/orca/data/ray_rdd.py:40]: good catch

File: pyzoo/zoo/orca/data/ray_rdd.py [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+        password = ray_ctx.redis_password
+        driver_ip = ray.services.get_node_ip_address()
+        uuid_str = str(uuid.uuid4())
+        meta_store = MetaStore.options(name=f"meta_store:{uuid_str}").remote()
+        resources = ray.cluster_resources()
+        nodes = []
+        for key, value in resources.items():
+            if key.startswith("node:"):
+                # if running in cluster, filter out driver ip
+                if not (not ray_ctx.is_local and key == f"node:{driver_ip}"):
```
Comment (jason-dai) [inline pyzoo/zoo/orca/data/ray_rdd.py:273]: if (ray_ctx.is_local or key != f"node:{driver_ip}")

File: pyzoo/zoo/orca/data/ray_rdd.py [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+        partition_ref = ray.put(partition)
+        ray.get(self.meta_store_handle.set_partition_ref.remote(idx, [partition_ref]))
+        return 0
+
+    def upload_multiple_partitions(self, idxs, partitions):
+        for idx, partition in zip(idxs, partitions):
+            self.upload_partition(idx, partition)
+        return 0
+
+    def upload_partitions_from_plasma(self, partition_id, plasma_object_id, object_store_address):
```
Comment (jason-dai) [inline pyzoo/zoo/orca/data/ray_rdd.py:61]: `upload_partitions_via_plasma`

File: pyzoo/zoo/orca/data/ray_rdd.py [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+        client = plasma.connect(object_store_address)
+        partition = client.get(plasma_object_id)
+        partition_ref = ray.put(partition)
+        ray.get(self.meta_store_handle.set_partition_ref(partition_id, [partition_ref]))
+        return 0
+
+
+def write_to_ray(idx, partition, redis_address, redis_password, partition_store_names):
+    partition = list(partition)
+    ray.init(address=redis_address, redis_password=redis_password, ignore_reinit_error=True)
```
Comment (jason-dai) [inline pyzoo/zoo/orca/data/ray_rdd.py:72]: have we tested it on 100s or 1000s of executors?

File: pyzoo/zoo/orca/data/ray_rdd.py [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+import random
+
+from zoo.ray import RayContext
+
+
+@ray.remote(num_cpus=0)
+class MetaStore:
+
+    def __init__(self):
+        self.partitions = {}
```
Comment (jason-dai) [inline pyzoo/zoo/orca/data/ray_rdd.py:27]: `partition_ids`?

File: pyzoo/zoo/orca/data/ray_rdd.py [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+
+    def set_partition_ref(self, idx, object_ref):
```
Comment (jason-dai) [inline pyzoo/zoo/orca/data/ray_rdd.py:29]: `set_partition_id`?

File: pyzoo/zoo/orca/data/ray_rdd.py [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+        self.partitions[idx] = object_ref[0]
+        return 0
+
+    def get_partition_ref(self, idx):
```
Comment (jason-dai) [inline pyzoo/zoo/orca/data/ray_rdd.py:33]: `get_partition_id`?

File: pyzoo/zoo/orca/data/ray_rdd.py [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+        return len(self.partitions)
+
+
+class PartitionUploader:
+
+    def __init__(self, meta_store_handle):
+        self.meta_store_handle = meta_store_handle
+
+    def upload_partition(self, idx, partition):
+        partition_ref = ray.put(partition)
```
Comment (jason-dai) [inline pyzoo/zoo/orca/data/ray_rdd.py:52]: `partition_id`

**Pull request merged**